lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs

```diff
@@ -533,6 +546,102 @@ spec = do
                 (#balance . #available)
                 (`shouldBe` Quantity (faucetAmt - feeEstMax - amt)) ra2
 
+    it "TRANS_CREATE_10 - Transaction with metadata" $ \ctx -> do
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = (1 :: Natural)
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = [json|{
+                "1": "hello"
+            }|]
+        let expected = TxMetadata (MD.MetaData (Map.singleton 1 (MD.S "hello")))
+        let payload = addTxMetadata txMeta basePayload
+
+        ra <- request @(ApiTransaction n) ctx
+            (Link.createTransaction @'Shelley wa) Default payload
+
+        verify ra
+            [ expectSuccess
+            , expectResponseCode HTTP.status202
+            , expectField (#status . #getApiT) (`shouldBe` Pending)
+            , expectField
+                (#metadata . #getApiTxMetadata)
+                (`shouldBe` Just (ApiT expected))
+            ]
+
+        eventually "metadata is confirmed in transaction list" $ do
+            let link = Link.listTransactions @'Shelley wa
+            rb <- request @([ApiTransaction n]) ctx link Default Empty
+            verify rb
+                [ expectResponseCode HTTP.status200
+                , expectListField 0 (#status . #getApiT) (`shouldBe` InLedger)
+                , expectListField 0
+                    (#metadata . #getApiTxMetadata)
+                    (`shouldBe` Just (ApiT expected))
+                ]
+
+    it "TRANS_CREATE_11 - Transaction with invalid metadata" $ \ctx -> do
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = (1 :: Natural)
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = Aeson.object ["1" .= T.replicate 65 "a"]
+        let payload = addTxMetadata txMeta basePayload
+
+        r <- request @(ApiTransaction n) ctx
+            (Link.createTransaction @'Shelley wa) Default payload
+
+        expectResponseCode @IO HTTP.status400 r
+        expectErrorMessage "fixme: api validation error message" r
+
+    it "TRANS_CREATE_12 - Transaction with too much metadata" $ \ctx -> do
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = (1 :: Natural)
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        -- This will encode to at least 64k of CBOR.
+        let txMeta = Aeson.object
+                [ (toText @Int i, Aeson.String (T.replicate 64 "a"))
+                | i <- [0..1023] ]
+        let payload = addTxMetadata txMeta basePayload
+
+        r <- request @(ApiTransaction n) ctx
+            (Link.createTransaction @'Shelley wa) Default payload
+
+        expectResponseCode @IO HTTP.status400 r
+        expectErrorMessage "fixme: transaction rejected message" r
```

Comment:
so does this one

Reply:
See latest update of the branch.
Turns out the error message for oversized transactions is not very good.
I think we should fix the error message - but in another PR.

Reply:
Too late. I've done it in this PR :grimacing:

Reply:
Thanks